pyuavcan/transport/can/media/pythoncan/_pythoncan.py

```diff
@@ -4,54 +4,343 @@
 # Author: Pavel Kirienko <pavel.kirienko@zubax.com>
 #
 
+import enum
+import time
+import errno
 import typing
-import logging
+import struct
 import asyncio
+import concurrent.futures
+import logging
+import threading
+import can  # type: ignore
+import pyuavcan.transport
 import pyuavcan.transport.can.media as _media
+import dataclasses
+import collections
 
 
 _logger = logging.getLogger(__name__)
 
 
+@dataclasses.dataclass(frozen=True)
+class _InterfaceParameters:
+    interface_name: str
+    channel_name:   str
+
+
+@dataclasses.dataclass(frozen=True)
+class _ClassicInterfaceParameters(_InterfaceParameters):
+    bitrate: int
+
+
+@dataclasses.dataclass(frozen=True)
+class _FDInterfaceParameters(_InterfaceParameters):
+    bitrate: typing.Tuple[int, int]
+
+
+def _construct_socketcan(parameters: _InterfaceParameters) -> can.ThreadSafeBus:
+    if isinstance(parameters, _ClassicInterfaceParameters):
+        return can.ThreadSafeBus(interface=parameters.interface_name,
+                                 channel=parameters.channel_name,
+                                 fd=False)    
+    elif isinstance(parameters, _FDInterfaceParameters):
+        return can.ThreadSafeBus(interface=parameters.interface_name,
+                                 channel=parameters.channel_name,
+                                 fd=True)    
+    else:
+        raise TypeError(f'Invalid parameters: {parameters}')   
+
+
+def _construct_kvaser(parameters: _InterfaceParameters) -> can.ThreadSafeBus:
+    if isinstance(parameters, _ClassicInterfaceParameters):
+        return can.ThreadSafeBus(interface=parameters.interface_name,
+                                 channel=parameters.channel_name,
+                                 bitrate=parameters.bitrate,
+                                 fd=False)
+    elif isinstance(parameters, _FDInterfaceParameters):
+        return can.ThreadSafeBus(interface=parameters.interface_name,
+                                 channel=parameters.channel_name,
+                                 bitrate=parameters.bitrate[0],
+                                 fd=True,
+                                 data_bitrate=parameters.bitrate[1])
+    else:
+        raise TypeError(f'Invalid parameters: {parameters}')
+
+
+def _construct_slcan(parameters: _InterfaceParameters) -> can.ThreadSafeBus:
+    if isinstance(parameters, _ClassicInterfaceParameters):
+        # only default ttyBaudrate is possible (115200)
+        return can.ThreadSafeBus(interface=parameters.interface_name,
+                                 channel=parameters.channel_name,
+                                 bitrate=parameters.bitrate)
+    elif isinstance(parameters, _FDInterfaceParameters):
+        raise TypeError('Interface does not support CAN FD: {}'.format(parameters.interface_name))
```

Comment:
This is minor but if you could update other occurrences to use f-strings that would be great. F-strings are more reliable because you can't accidentally mess up the formatting.

```suggestion
        raise TypeError(f'Interface does not support CAN FD: {parameters.interface_name}')
```